src/librustdoc/clean/types.rs

```diff
@@ -1299,13 +1305,21 @@ crate enum TypeKind {
 }
 
 crate trait GetDefId {
+    /// Doesn't retrieve primitive types `DefId`. Use `def_id_full` if you want it.
     fn def_id(&self) -> Option<DefId>;
+    /// Retrieves all types' `DefId` (including primitives). If you're not interested about
+    /// primitives, use `def_id`.
+    fn def_id_full(&self, cache: &Cache) -> Option<DefId>;
```

Comment:
Could you elaborate this documentation a bit? As someone who doesn't know much about `Cache`, I don't quite understand the difference between these two.

Reply:
Then I'll need your help to reword it! :p

Basically, to get primitive types' `DefId` we need the `Cache` type (because that's where they're stored). So unless you want the `DefId` of a primitive type, you can use `def_id`, which doesn't need the `Cache` parameter.

Reply:
What kind of items would you call `def_id_full` on? Just `clean::Type`? Or anything that contains a primitive type somewhere?

Reply:
The `clean` types (so `clean::Type` too).

Reply:
You said:

> So unless you want the `DefId` of a primitive type, you can use `def_id`, which doesn't need the `Cache` parameter.

What do you mean by "unless you want the `DefId` of a primitive type"? Does that mean that you need `def_id_full` whenever there's a primitive type inside of whatever you're calling `def_id_full` on?

Reply:
Yes, exactly. If you have a `clean::Type` which is containing a primitive type for example, then `def_id` will return `None` but `def_id_full` will return the primitive type `DefId`.

Reply:
Okay, then I would suggest something like this for the docs:

```suggestion
    /// Use this method to get the [`DefId`] of a [`clean`] AST node.
    /// This will return [`None`] when called on a primitive [`clean::Type`].
    /// Use [`Self::def_id_full`] if you are calling it on a primitive [`clean::Type`].
    ///
    /// [`clean`]: crate::clean
    /// [`clean::Type`]: Type
    fn def_id(&self) -> Option<DefId>;
    /// Use this method to get the [`DefId`] of a [`clean`] AST node that may be
    /// a primitive [`clean::Type`].
    ///
    /// See [`Self::def_id`] for more.
    ///
    /// [`clean::Type`]: Type
    fn def_id_full(&self, cache: &Cache) -> Option<DefId>;
```

(Assuming what I put in the docs is correct.)

Reply:
I'm not a big fan of your suggestion. I find it more confusing than mine.

Reply:
What's more confusing about it? I find *yours* more confusing ;)

Reply:
Oh well, let's go with your suggestion then. I'm the one who made the implementation so might not be the best to judge here haha.